### PR TITLE
drivers/periph/spi: Implement thread safety for all SPI devices

### PR DESCRIPTION
**Pull request by gebart**
This implements #2314 for the SPI drivers. (edit: Except for STM32F4 which is implemented in #2290)
I added the generic implementation of SPI thread safety as proposed by @haukepetersen in #2290 to all CPU `periph/spi` implementations and made all drivers which are currently using `periph/spi` to acquire/release the bus when necessary.

I also found a mistake in the nrf24l01p driver, the CS line was cleared/set twice when turning on/off the device.

### CODE REVIEW

File: drivers/nrf24l01p/nrf24l01p.c [hunk 1]
```diff
@@ -189,12 +199,10 @@ int nrf24l01p_on(nrf24l01p_t *dev)
     char read;
     int status;
 
-    gpio_clear(dev->cs);
     nrf24l01p_read_reg(dev, REG_CONFIG, &read);
     hwtimer_spin(DELAY_CS_TOGGLE_TICKS);
```
Comment (gebart) [inline drivers/nrf24l01p/nrf24l01p.c]: Are these delays needed, or are they accidental left-overs from a previous refactoring?

File: drivers/nrf24l01p/nrf24l01p.c [hunk 2]
```diff
@@ -189,12 +199,10 @@ int nrf24l01p_on(nrf24l01p_t *dev)
...
     status = nrf24l01p_write_reg(dev, REG_CONFIG, (read | PWR_UP));
     hwtimer_spin(DELAY_CS_TOGGLE_TICKS);
```
Comment (gebart) [inline drivers/nrf24l01p/nrf24l01p.c]: Are these delays needed, or are they accidental left-overs from a previous refactoring?

File: drivers/nrf24l01p/nrf24l01p.c [hunk 3]
```diff
@@ -206,12 +214,10 @@ int nrf24l01p_off(nrf24l01p_t *dev)
     char read;
     int status;
 
-    gpio_clear(dev->cs);
     nrf24l01p_read_reg(dev, REG_CONFIG, &read);
     hwtimer_spin(DELAY_CS_TOGGLE_TICKS);
     status = nrf24l01p_write_reg(dev, REG_CONFIG, (read & ~PWR_UP));
     hwtimer_spin(DELAY_CS_TOGGLE_TICKS);
```
Comment (gebart) [inline drivers/nrf24l01p/nrf24l01p.c]: Are these delays needed, or are they accidental left-overs from a previous refactoring?

**Comment by gebart** [comment]
@haukepetersen I added you as author on the affected `$CPU/periph/spi.c` files since you made the actual implementation work.

File: drivers/nrf24l01p/nrf24l01p.c (hunk 1 above)
Comment (PeterKietzmann) [inline drivers/nrf24l01p/nrf24l01p.c]: @gebart, I assume they are not. They have been introduced in #1704 in cause of "inexplicable" timing errors. Just today we found one possible reason for this in #2315. I will test this soon but I have not time for this today.

**Comment by PeterKietzmann** [comment]
BTW: thanks for doing all this 	task requiring great diligence!

**Comment by gebart** [comment]
Updated according to https://github.com/RIOT-OS/RIOT/pull/2290#issuecomment-70249763

File: drivers/nrf24l01p/nrf24l01p.c [hunk 4]
```diff
@@ -231,12 +239,16 @@ int nrf24l01p_read_payload(nrf24l01p_t *dev, char *answer, unsigned int size)
...
+    /* Acquire exclusive access to the bus. */
+    spi_acquire(dev->spi);
     gpio_clear(dev->cs);
     hwtimer_spin(DELAY_CS_TOGGLE_TICKS);
     status = spi_transfer_regs(dev->spi, CMD_R_RX_PAYLOAD, 0, answer, size);
     hwtimer_spin(DELAY_CS_TOGGLE_TICKS);
     gpio_set(dev->cs);
     hwtimer_spin(DELAY_AFTER_FUNC_TICKS);
+    /* Release the bus for other threads. */
+    spi_release(spi-dev);
```
Comment (PeterKietzmann) [inline drivers/nrf24l01p/nrf24l01p.c]: missing "`>`"

Comment (PeterKietzmann) [inline drivers/nrf24l01p/nrf24l01p.c]: it must be `spi_release(dev->spi);`

File: drivers/nrf24l01p/nrf24l01p.c (hunk 1 above)
Comment (PeterKietzmann) [inline drivers/nrf24l01p/nrf24l01p.c]: @gebart yes these delays are superfluous. Would be nice if you could remove them.

**Comment by PeterKietzmann** [comment]
@gebart: except from the small changes in `nrf24l01p.c` I would ACK this PR

**Comment by gebart** [comment]
@PeterKietzmann Fixed the typo according to your comment, squash and go? (once Travis is done)

Should I leave each individual component as a separate commit as now and only squash the small comment fixes or squash it all into one big commit?

**Comment by PeterKietzmann** [comment]
@gebart I would propose to squash the nrf24l01p-, cc110x- and at86rf231-commits but I'm not sure if it's necessary to keep all the other ones separately. Maybe @LudwigOrtmann ha an idea about that?

**Comment by gebart** [comment]
I'd prefer to leave it at one commit per cpu, just for history clarity.

**Comment by PeterKietzmann** [comment]
Yep that's what was in my mind too. I was just to stupid to phrase. But still, I think you can squash nrf24l01p-, cc110x- and at86rf231-commits. Would like to hear another opinion, maybe from Ludwig?

**Comment by LudwigOrtmann** [comment]
I completely agree with the atomicity of the commits not prefixed 'squash' ;)
However, I notice fluctuations in the naming of commits. Sometimes "Require" is used instead of "Acquire". Is that on purpose?

**Comment by gebart** [comment]
@LudwigOrtmann Require was probably just a typo on my side. I will rewrite the message during squashing.

**Comment by gebart** [comment]
I think the removing of the extra delays in nrf24l01p shall remain as a stand alone commit for easier bisection in any future regressions.

@PeterKietzmann ACK if you are fine with the current commit line

**Comment by PeterKietzmann** [comment]
I'm happy. ACK&go

**Pull request merged**